File: apps/worker/orbit_worker/scorecard.py
from __future__ import annotations

from .domain import DIMENSION_WEIGHTS, RECOMMENDATION_ORDER, RECOMMENDATION_RANK, SCORE_DIMENSIONS, clamp_score, dedupe_preserve_order, round_half_up
from .schemas import AgentReview, CanonicalPortfolio, ConflictRecord, DimensionScore, Finding, Scorecard, validate_scorecard


def mean(values: list[float]) -> float:
    return 0 if not values else sum(values) / len(values)


def fixed_2(value: float) -> float:
    return round_half_up(value, 2)


def downgrade_recommendation(recommendation: str) -> str:
    index = max(0, RECOMMENDATION_RANK[recommendation] - 1)
    return RECOMMENDATION_ORDER[index]


def calculate_aggregate_metrics(dimension_scores: list[dict[str, object] | DimensionScore]) -> tuple[float, float, float, list[str]]:
    normalized_scores = [
        entry if isinstance(entry, dict) else entry.model_dump(mode="python")
        for entry in dimension_scores
    ]
    weighted_composite_score = fixed_2(
        sum(float(entry["score"]) * DIMENSION_WEIGHTS[str(entry["dimension"])] for entry in normalized_scores) / 100,
    )
    average_confidence = fixed_2(mean([float(entry["confidence"]) for entry in normalized_scores]))
    average_evidence_completeness = fixed_2(mean([float(entry["evidence_completeness"]) for entry in normalized_scores]))
    severity_flags = sorted(
        dedupe_preserve_order(
            flag
            for entry in normalized_scores
            for flag in list(entry["severity_flags"])
        )
    )
    return weighted_composite_score, average_confidence, average_evidence_completeness, severity_flags


def determine_final_recommendation(
    weighted_composite_score: float,
    average_confidence: float,
    average_evidence_completeness: float,
    findings: list[Finding],
    conflicts: list[ConflictRecord],
) -> tuple[str, bool]:
    critical_governance_finding = next(
        (
            finding
            for finding in findings
            if finding.severity == "critical" and finding.category in {"security_and_compliance", "ai_reliability"}
        ),
        None,
    )

    if weighted_composite_score >= 4.2 and average_confidence >= 0.7 and average_evidence_completeness >= 0.8 and not critical_governance_finding:
        final_recommendation = "Strong Proceed"
    elif weighted_composite_score >= 3.5 and not critical_governance_finding:
        final_recommendation = "Proceed with Conditions"
    elif weighted_composite_score >= 2.8:
        final_recommendation = "Pilot Only"
    elif weighted_composite_score >= 2.0:
        final_recommendation = "High Risk"
    else:
        final_recommendation = "Do Not Proceed"

    high_conflicts = len([conflict for conflict in conflicts if conflict.severity == "high"])
    if high_conflicts >= 4 and RECOMMENDATION_RANK[final_recommendation] > RECOMMENDATION_RANK["Pilot Only"]:
        final_recommendation = downgrade_recommendation(final_recommendation)
    if critical_governance_finding and RECOMMENDATION_RANK[final_recommendation] > RECOMMENDATION_RANK["High Risk"]:
        final_recommendation = "High Risk"

    return final_recommendation, bool(critical_governance_finding)


def build_scorecard_conditions(findings: list[Finding], conflicts: list[ConflictRecord]) -> list[str]:
    return dedupe_preserve_order(
        [
            *[
                finding.recommended_action
                for finding in findings
                if finding.severity in {"major", "critical"}
            ][:6],
            *[
                f"Resolve {conflict.conflict_type} on {conflict.topic} before broad rollout."
                for conflict in conflicts
                if conflict.severity == "high"
            ][:3],
        ]
    )[:8]


def build_committee_scorecard(
    portfolio: CanonicalPortfolio,
    run_id: str,
    reviews: list[AgentReview],
    conflicts: list[ConflictRecord],
) -> Scorecard:
    dimension_scores = []
    for dimension in SCORE_DIMENSIONS:
        entries = [score for review in reviews for score in review.dimension_scores if score.dimension == dimension]
        dimension_scores.append(
            {
                "dimension": dimension,
                "score": clamp_score(mean([entry.score for entry in entries])),
                "confidence": fixed_2(mean([entry.confidence for entry in entries])),
                "evidence_completeness": fixed_2(mean([entry.evidence_completeness for entry in entries])),
                "severity_flags": sorted(dedupe_preserve_order(flag for entry in entries for flag in entry.severity_flags)),
                "rationale": f"Committee score derived from {len(entries)} structured reviewer dimension entries.",
                "evidence_refs": dedupe_preserve_order(ref for entry in entries for ref in entry.evidence_refs),
            }
        )

    weighted_composite_score, average_confidence, average_evidence_completeness, severity_flags = calculate_aggregate_metrics(
        dimension_scores
    )
    findings = [finding for review in reviews for finding in review.findings]
    final_recommendation, override_applied = determine_final_recommendation(
        weighted_composite_score,
        average_confidence,
        average_evidence_completeness,
        findings,
        conflicts,
    )
    conditions = build_scorecard_conditions(findings, conflicts)

    return validate_scorecard(
        {
            "portfolio_id": portfolio.portfolio_id,
            "run_id": run_id,
            "dimension_scores": dimension_scores,
            "weighted_composite_score": weighted_composite_score,
            "average_confidence": average_confidence,
            "average_evidence_completeness": average_evidence_completeness,
            "severity_flags": severity_flags,
            "final_recommendation": final_recommendation,
            "override_applied": override_applied,
            "conditions": conditions,
        }
    )

File: archive/js-baseline/packages/orbit-scoring/src/scorecard.js
// @ts-check
'use strict';

const { DIMENSION_WEIGHTS, RECOMMENDATION_ORDER, RECOMMENDATION_RANK, SCORE_DIMENSIONS, clampScore } = require('../../orbit-core/src/domain');
const { validateScorecard } = require('../../orbit-core/src/schema');

function mean(values) {
  return values.length === 0 ? 0 : values.reduce((sum, value) => sum + value, 0) / values.length;
}

function downgradeRecommendation(recommendation) {
  const index = Math.max(0, RECOMMENDATION_RANK[recommendation] - 1);
  return RECOMMENDATION_ORDER[index];
}

function buildCommitteeScorecard(portfolio, runId, reviews, conflicts) {
  const dimensionScores = SCORE_DIMENSIONS.map((dimension) => {
    const entries = reviews.flatMap((review) => review.dimension_scores.filter((score) => score.dimension === dimension));
    const score = clampScore(mean(entries.map((entry) => entry.score)));
    const confidence = Number(mean(entries.map((entry) => entry.confidence)).toFixed(2));
    const evidenceCompleteness = Number(mean(entries.map((entry) => entry.evidence_completeness)).toFixed(2));
    const severityFlags = Array.from(new Set(entries.flatMap((entry) => entry.severity_flags))).sort();
    return {
      dimension,
      score,
      confidence,
      evidence_completeness: evidenceCompleteness,
      severity_flags: severityFlags,
      rationale: `Committee score derived from ${entries.length} structured reviewer dimension entries.`,
      evidence_refs: Array.from(new Set(entries.flatMap((entry) => entry.evidence_refs))),
    };
  });

  const weightedCompositeScore = Number((dimensionScores.reduce((sum, entry) => sum + (entry.score * DIMENSION_WEIGHTS[entry.dimension]), 0) / 100).toFixed(2));
  const averageConfidence = Number(mean(dimensionScores.map((entry) => entry.confidence)).toFixed(2));
  const averageEvidenceCompleteness = Number(mean(dimensionScores.map((entry) => entry.evidence_completeness)).toFixed(2));
  const severityFlags = Array.from(new Set(reviews.flatMap((review) => review.dimension_scores.flatMap((entry) => entry.severity_flags)))).sort();
  const findings = reviews.flatMap((review) => review.findings);
  const criticalGovernanceFinding = findings.find((finding) => finding.severity === 'critical' && ['security_and_compliance', 'ai_reliability'].includes(finding.category));

  let finalRecommendation;
  if (weightedCompositeScore >= 4.2 && averageConfidence >= 0.7 && averageEvidenceCompleteness >= 0.8 && !criticalGovernanceFinding) {
    finalRecommendation = 'Strong Proceed';
  } else if (weightedCompositeScore >= 3.5 && !criticalGovernanceFinding) {
    finalRecommendation = 'Proceed with Conditions';
  } else if (weightedCompositeScore >= 2.8) {
    finalRecommendation = 'Pilot Only';
  } else if (weightedCompositeScore >= 2.0) {
    finalRecommendation = 'High Risk';
  } else {
    finalRecommendation = 'Do Not Proceed';
  }

  const highConflicts = conflicts.filter((conflict) => conflict.severity === 'high').length;
  if (highConflicts >= 4 && RECOMMENDATION_RANK[finalRecommendation] > RECOMMENDATION_RANK['Pilot Only']) {
    finalRecommendation = downgradeRecommendation(finalRecommendation);
  }
  if (criticalGovernanceFinding && RECOMMENDATION_RANK[finalRecommendation] > RECOMMENDATION_RANK['High Risk']) {
    finalRecommendation = 'High Risk';
  }

  const conditions = Array.from(new Set([
    ...findings.filter((finding) => ['major', 'critical'].includes(finding.severity)).slice(0, 6).map((finding) => finding.recommended_action),
    ...conflicts.filter((conflict) => conflict.severity === 'high').slice(0, 3).map((conflict) => `Resolve ${conflict.conflict_type} on ${conflict.topic} before broad rollout.`),
  ])).slice(0, 8);

  return validateScorecard({
    portfolio_id: portfolio.portfolio_id,
    run_id: runId,
    dimension_scores: dimensionScores,
    weighted_composite_score: weightedCompositeScore,
    average_confidence: averageConfidence,
    average_evidence_completeness: averageEvidenceCompleteness,
    severity_flags: severityFlags,
    final_recommendation: finalRecommendation,
    override_applied: Boolean(criticalGovernanceFinding),
    conditions,
  });
}

module.exports = {
  buildCommitteeScorecard,
};
